File: leetcode/lc-us/cpp/974.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int subarraysDivByK(vector<int> &nums, int k) {
        int cnt = 0;
        unordered_map<int, int> map;
        map[0] = 1;
        for (int i = 0, sum = 0; i < nums.size(); i++) {
            sum += nums[i];
            int r = sum % k;
            if (r < 0) r += k;
            if (map.count(r)) {
                cnt += map[r];
            }
            map[r]++;
        }
        return cnt;
    }
};

File: leetcode/lc-us/py/974.py
from collections import defaultdict

from typing import List


class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        m = defaultdict(int)
        m[0] = 1
        s = res = 0
        for num in nums:
            s = (s + num) % k
            res += m[s]
            m[s] += 1
        return res
